### PR TITLE
[core] Shard tensor-parallel checkpoints on load and save

### LINKED ISSUE #14533: better reporting of errors when using TP
I think the following aren't implemented at the moment (which is fine; just flagging). 

* **Sharded loading**. from_pretrained should stream shards straight to each rank's DTensor rather than materializing the full checkpoint then slicing — otherwise TP saves you nothing at load time. And check `save_pretrained` / state_dict calls `.full_tensor()` or uses DCP. I think we should at least raise when `save_pretrained()` is called in case TP is enabled?
* **LoRA loading**. for a colwise base layer, lora_A replicated + lora_B colwise; for rowwise, lora_A rowwise + lora_B replicated. If the plan doesn't cover PEFT layers, loading an adapter onto a TP model will either error or be wrong. I think we should detect if the model has `peft` layers injected and raise if TP is requested?
* **Quantization, offloading**. We should probably also raise when these are requested?

_Originally posted by @sayakpaul in https://github.com/huggingface/diffusers/pull/13718#discussion_r3662821938_

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

Fixes #14533 

This is a follow-up of the Tensor Parallelism support in #13781, TP previously required loading the whole checkpoint on every rank and resharding it afterwards, so per-rank memory was the full model size. In this PR, we adapt the shard loading (`.from_pretrained()`) and saving (`.save_pretrained()`) to be tp-aware:

- Load: `from_pretrained(..., parallel_config=...)` shards while reading, each rank slices only its own part of every `_tp_plan` weight off disk, straight into a DTensor on its device. Unsupported combinations: device_map / quantization / use_flashpack / DDUF / non-safetensors -> raise.
- Save: `save_pretrained()` gathers the shards back to a normal checkpoint
    - collective: all on all ranks, rank 0 writes 
    - save_pretrained(..., dcp=True) writes per-rank

```python
import torch
import torch.distributed as dist

from diffusers import Flux2Transformer2DModel, TensorParallelConfig

dist.init_process_group()                                  # "nccl" on CUDA, "neuron" on Trainium
tp = TensorParallelConfig(tp_degree=dist.get_world_size())

# LOAD: shard while reading, never materialize the full checkpoint ----
model = Flux2Transformer2DModel.from_pretrained(
    "black-forest-labs/FLUX.2-klein-9B",
    subfolder="transformer",
    torch_dtype=torch.bfloat16,
    parallel_config=tp,
)

# SAVE: gather back
model.save_pretrained("out/full")

# SAVE (b): sharded distributed checkpoint, nothing gathered
# model.save_pretrained("out/sharded", dcp=True)

# LOAD BACK a distributed ckpt
# model = Flux2Transformer2DModel.from_pretrained("out/sharded", parallel_config=tp)
```

Besides above:
- `_check_tp_model_state`, called from `apply_tensor_parallel`, rejects a model that is quantized, group-offloaded, placed by accelerate (`device_map` or CPU offload), or has PEFT layers injected.
- `load_lora_adapter` refuses a tensor-parallel model.
- `save_pretrained` refuses a quantized TP model.

## Before submitting
- [x] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR? If so:
  - [x] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [x] Did you run the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on the diff?
  - [x] Did you share the final self-review notes in the PR description or a comment?
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [x] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)? (important for complex PRs)
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?
- [x] Are you the author (or part of the team) of the model/pipeline (only applicable for model/pipeline related PRs)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/en/training/distributed_inference.md [hunk 1]
```diff
@@ -436,43 +436,42 @@ pipeline = DiffusionPipeline.from_pretrained(
...
 
-Pass a [`TensorParallelConfig`] to [`~ModelMixin.enable_parallelism`]. `tp_degree` is the number of devices to shard across and must divide the model's number of attention heads. The model must define a `_tp_plan` (a flat mapping of module-name globs to a `"colwise"`/`"rowwise"` style).
+Pass a [`TensorParallelConfig`] to the `parallel_config` argument of the model's [`~ModelMixin.from_pretrained`]. `tp_degree` is the number of devices to shard across and must divide the model's number of attention heads. The model must define a `_tp_plan` (a flat mapping of module-name globs to a `"colwise"`/`"rowwise"` style).
+
+Loading this way shards the checkpoint *while reading it*: each rank reads only its own slice of each sharded weight and places it straight onto its own device. Nothing full-size is ever materialized, so per-rank memory falls as `tp_degree` rises.
```
Comment (sayakpaul) [inline docs/source/en/training/distributed_inference.md:441]: Oh really! This is very cool. Could we also present a small comparison between the loading time with and without this way of loading?

File: docs/source/en/training/distributed_inference.md [hunk 2]
```diff
@@ -484,6 +483,25 @@ torchrun --nproc-per-node 4 tensor_parallel_flux.py
 
 `tp_degree` is taken from `world_size` above, so `--nproc-per-node 4` shards the transformer across 4 devices.
 
+A tensor-parallel `parallel_config` cannot be combined with `device_map`, `quantization_config`, `low_cpu_mem_usage=False`, `use_flashpack=True`, or non-safetensors weights; each raises rather than quietly falling back to loading the full checkpoint. Tensor parallelism also cannot be combined with quantization, offloading, or LoRA adapters at all — the parameters it shards have to be plain parameters owned by the model — so those raise however the model is sharded. To shard a model that is already in memory, call [`~ModelMixin.enable_parallelism`] with the same config instead — that loads everything first and reshards it, so it costs full checkpoint memory on every rank.
```
Comment (sayakpaul) [inline docs/source/en/training/distributed_inference.md:486]: Interesting that we cannot load TP with quantization. Do we know why?

File: docs/source/en/training/distributed_inference.md [hunk 3]
```diff
@@ -484,6 +483,25 @@ torchrun --nproc-per-node 4 tensor_parallel_flux.py
...
+
+### Saving a tensor-parallel model
+
+[`~ModelMixin.save_pretrained`] gathers the shards back into ordinary full tensors, so the result is a normal checkpoint that loads with or without tensor parallelism. Gathering is a collective, so call it on **every** rank; only rank 0 writes.
```
Comment (sayakpaul) [inline docs/source/en/training/distributed_inference.md:490]: That is cool! However, do we have to ship this yet? I don't have any strong opinions. @DN6 WDYT?

File: docs/source/en/training/distributed_inference.md [hunk 4]
```diff
@@ -484,6 +483,25 @@ torchrun --nproc-per-node 4 tensor_parallel_flux.py
...
+pipeline.transformer.save_pretrained("flux2-transformer")
+```
+
+For a model too large to gather onto a single rank, pass `dcp=True` to write a [distributed checkpoint](https://pytorch.org/docs/stable/distributed.checkpoint.html) instead. Every rank writes its own shards, so no full tensor is ever formed.
+
+```py
+pipeline.transformer.save_pretrained("flux2-transformer-dcp", dcp=True)
+```
+
+`from_pretrained` detects such a directory automatically, and reads it back with the same `parallel_config` you saved it under. Because a packed projection's shards are stored interleaved by the writing degree, the checkpoint only loads at that same `tp_degree`, and only with tensor parallelism — anything else raises rather than silently returning wrong weights. It is also local-only: a distributed checkpoint is recognized by the `.metadata` file in its directory, so it cannot be pushed to or loaded from the Hub. To lift any of these restrictions, re-save with the default (gathered) path, which produces an ordinary checkpoint.
```
Comment (sayakpaul) [inline docs/source/en/training/distributed_inference.md:503]: Not sure if we want to ship this right away because it is introducing quite a bit of code changes for something we don't know to be impactful yet.

File: src/diffusers/hooks/tensor_parallel.py [hunk 5]
```diff
@@ -123,32 +269,24 @@ def _make_packed_col(marker: PackedColwiseParallel) -> ColwiseParallel:
...
-                                f"{bs} is not divisible by {tp_size}."
-                            )
-                        chunk = bs // tp_size
-                        parts.append(full[offset + rank * chunk : offset + (rank + 1) * chunk].contiguous())
-                        offset += bs
-                    local = torch.cat(parts, dim=0)
-                    dist_param = nn.Parameter(
-                        DTensor.from_local(local, device_mesh, [Shard(0)], run_check=False),
-                        requires_grad=param.requires_grad,
+                    local = _local_shard(full, 0, _blocks_to_block_sizes(full.shape[0], blocks), device_mesh)
```
Comment (sayakpaul) [inline src/diffusers/hooks/tensor_parallel.py:282]: This reads a bit difficult. Can we break them further a bit?

File: src/diffusers/hooks/tensor_parallel.py [hunk 6]
```diff
@@ -240,16 +366,121 @@ def _partition_linear_fn(self, name, module, device_mesh):
...
     if tp_mesh is None:
         raise ValueError("`config._mesh` is None. Call `config.setup(rank, world_size, device)` before applying TP.")
 
+    num_heads = getattr(model.config, "num_attention_heads", None)
+    if num_heads is not None and num_heads % config._tp_degree != 0:
+        raise ValueError(f"`tp_degree` ({config._tp_degree}) must divide the number of attention heads ({num_heads}).")
+
+    # Before the device-type check below, so that a quantized or offloaded model reports what is actually wrong with
+    # it rather than being turned away for its device type.
+    _check_tp_model_state(model)
```
Comment (sayakpaul) [inline src/diffusers/hooks/tensor_parallel.py:482]: Nice!

**Review by sayakpaul — COMMENTED** [review]
Left some high-level comments. 

My main comment is if we want to ship the advanced features of rank aware save and load yet. I am leaning towards raising when we encouter those situations and simplify the code a bit. This way, we can see if the community wants this feature and ship it when we have enough interest. But I would like to double-check with @DN6 on this too.

File: src/diffusers/hooks/tensor_parallel.py (hunk 6 above)
Comment (sayakpaul) [reply src/diffusers/hooks/tensor_parallel.py:482]: But should we rather have them in `enable_parallelism()` like the validations performed by CP?

File: src/diffusers/models/model_loading_utils.py [hunk 7]
```diff
@@ -381,6 +381,99 @@ def _load_shard_file(
     return offload_index, state_dict_index, mismatched_keys, error_msgs
 
 
+def _load_shard_file_tp(
```
Comment (sayakpaul) [inline src/diffusers/models/model_loading_utils.py:384]: Pausing the review on these since they are related to save/load which we're still deciding on.

Comment (DN6) [inline src/diffusers/models/model_loading_utils.py]: Small nit. _load_shard_file_with_tp

File: src/diffusers/hooks/tensor_parallel.py (hunk 6 above)
Comment (JingyaHuang) [reply src/diffusers/hooks/tensor_parallel.py]: yeah sure.

File: docs/source/en/training/distributed_inference.md (hunk 1 above)
Comment (JingyaHuang) [reply docs/source/en/training/distributed_inference.md]: Here's a quick benchmark:

Synthetic checkpoint (FLUX.1-shaped, 1.33B params, 2.65GB bf16):

| mode | tp_degree | load time | peak GPU/rank | peak CPU/rank |
|---|---|---|---|---|
| streamed (`from_pretrained(parallel_config=...)`) | 2 | 1.92s ± 0.11 | 1.71GB | 2.70GB |
| materialize+reshard (`from_pretrained` + `enable_parallelism`) | 2 | 3.06s ± 0.02 | 1.69GB | 4.08GB |
| streamed (`from_pretrained(parallel_config=...)`) | 4 | 1.45s ± 0.03 | 1.23GB | 2.20GB |
| materialize+reshard (`from_pretrained` + `enable_parallelism`) | 4 | 3.29s ± 0.02 | 1.23GB | 4.08GB |

- Speed: streaming is ~37% faster at tp_degree=2, ~56% faster at tp_degree=4 on the synthetic checkpoint (1.92s vs 3.06s, 1.45s vs 3.29s). 
- GPU memory: basically the same between the two.
- CPU/host memory: 34% less at tp_degree=2, 46% less at tp_degree=4 (2.70GB vs 4.08GB, 2.20GB vs 4.08GB).

File: docs/source/en/training/distributed_inference.md (hunk 2 above)
Comment (JingyaHuang) [reply docs/source/en/training/distributed_inference.md]: It's mostly about scoping... I didn't want to handle "quantization + TP" sharding in the scale of this PR. IMO, it could be possible if we shard and then quantize, and then according to the quantization configs and the backend, some might work, some doesn't... I would rather raise for now, we could consider supporting the combo properly if the community shows interest.
